Using an interface with a custom `str` method in a struct produces a C error

### Describe the bug

When an interface with a custom `str` method is used in a struct and that struct is printed a C error is produced.

```v
module main

interface Context {
	str() string
}

struct OtherContext {
	a string
}

pub fn (o &OtherContext) str() string {
	return 'OtherContext'
}

struct WithContext {
pub mut:
	ctx Context
}

fn with_function(ctx Context) {
	println(ctx)
}

fn main() {
	mut s := WithContext{
		ctx: OtherContext{}
	}
	println(s)

	with_function(OtherContext{})
}
```

### Expected Behavior

Output:
```
OtherContext
OtherContext
```


### Current Behavior

```
/tmp/v_60000/code.12154154399512805420.tmp.c:2208: warning: implicit declaration of function 'main__Context_str'
/tmp/v_60000/code.12154154399512805420.tmp.c:2208: error: '{' expected (got ";")
builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps

Run the code above

### Possible Solution


To get rid of this error you can implement the `str` method on the `Context` interface, but then the `str` method of `OtherContext` is ignored.

```v
interface Context {
// 	str() string
}

pub fn (c &Context) str() string {
	return 'Context'
}
```

Output:
```
WithContext{
    ctx: Context
}
Context
```

it is also possible to circumvent this error by matching the interface type in it's `str` method.

```v
pub fn (c &Context) str() string {
	match c {
		OtherContext {
			return c.str()
		}
		else {
			return 'Context'
		}
	}
}
```

Output:
```
WithContext{
    ctx: &OtherContext
}
&OtherContext
```

But this code automatically puts a `&` character before the type output. And you would have to know every struct that implements `Context` in advance.

Not to mention that this solution seems redundant and should be intergrated with the compiler.

### Additional Information/Context

_No response_

### V version

0.3.4

### Environment details (OS name and version, etc.)

V full version: V 0.3.4 ae8e9af.752e4c2
OS: linux, Kali GNU/Linux Rolling (WSL 2)

CC version: cc (Debian 12.2.0-14) 12.2.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -1068,6 +1068,12 @@ fn struct_auto_str_func(sym &ast.TypeSymbol, lang ast.Language, _field_type ast.
 	} else if _field_type.has_flag(.option) || should_use_indent_func(sym.kind) {
 		obj := '${deref}it.${final_field_name}${sufix}'
 		if has_custom_str {
+			if sym.kind == .interface_ {
+				mut s := '${fn_name.trim_string_right('_str')}_name_table[${obj}'
+				dot := if field_type.is_ptr() { '->' } else { '.' }
+				s += '${dot}_typ]._method_str(${obj}${dot}_object)'
+				return s, true
+			}
 			return '${fn_name}(${obj})', true
 		}
 		return 'indent_${fn_name}(${obj}, indent_count + 1)', true

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -1068,7 +1068,7 @@ fn struct_auto_str_func(sym &ast.TypeSymbol, lang ast.Language, _field_type ast.
 	} else if _field_type.has_flag(.option) || should_use_indent_func(sym.kind) {
 		obj := '${deref}it.${final_field_name}${sufix}'
 		if has_custom_str {
-			if sym.kind == .interface_ {
+			if sym.kind == .interface_ && (sym.info as ast.Interface).defines_method('str') {
 				iface_obj := 'it.${final_field_name}${sufix}'
 				dot := if field_type.is_ptr() { '->' } else { '.' }
 				return '${fn_name.trim_string_right('_str')}_name_table[${iface_obj}${dot}_typ]._method_str(${iface_obj}${dot}_object)', true

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -1069,10 +1069,9 @@ fn struct_auto_str_func(sym &ast.TypeSymbol, lang ast.Language, _field_type ast.
 		obj := '${deref}it.${final_field_name}${sufix}'
 		if has_custom_str {
 			if sym.kind == .interface_ {
-				mut s := '${fn_name.trim_string_right('_str')}_name_table[${obj}'
+				iface_obj := 'it.${final_field_name}${sufix}'
 				dot := if field_type.is_ptr() { '->' } else { '.' }
-				s += '${dot}_typ]._method_str(${obj}${dot}_object)'
-				return s, true
+				return '${fn_name.trim_string_right('_str')}_name_table[${iface_obj}${dot}_typ]._method_str(${iface_obj}${dot}_object)', true
 			}
 			return '${fn_name}(${obj})', true
 		}

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -108,6 +108,7 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 		unwrap_option := expr is ast.Ident && expr.or_expr.kind == .propagate_option
 		exp_typ := if unwrap_option { typ.clear_flag(.option) } else { typ }
 		is_ptr := exp_typ.is_ptr()
+		is_dump_expr := expr is ast.DumpExpr
 		is_var_mut := expr.is_auto_deref_var()
 		str_fn_name := g.get_str_fn(exp_typ)
 		if is_ptr && !is_var_mut {
@@ -130,7 +131,14 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 		}
 		g.write('${str_fn_name}(')
 		if str_method_expects_ptr && !is_ptr {
-			g.write('&')
+			if is_dump_expr {
+				g.write('ADDR(${g.typ(typ)}, ')
+				defer {
+					g.write(')')
+				}
+			} else {
+				g.write('&')
+			}
 		} else if is_ptr && typ.has_flag(.option) {
 			g.write('*(${g.typ(typ)}*)&')
 		} else if !str_method_expects_ptr && !is_shared && (is_ptr || is_var_mut) {

--- a/vlib/v/tests/interface_auto_str_test.v
+++ b/vlib/v/tests/interface_auto_str_test.v
@@ -1,0 +1,31 @@
+interface Context {
+	str() string
+}
+
+struct OtherContext {
+	a string
+}
+
+pub fn (o &OtherContext) str() string {
+	return 'OtherContext'
+}
+
+struct WithContext {
+pub mut:
+	ctx Context
+}
+
+fn with_function(ctx Context) {
+	println(ctx)
+}
+
+fn test_main() {
+	mut s := WithContext{
+		ctx: OtherContext{}
+	}
+	assert dump(s) == WithContext{
+		ctx: OtherContext{}
+	}
+
+	assert dump(OtherContext{}) == OtherContext{}
+}

